### ISSUE TITLE
Implement VPN termination

### ISSUE THREAD

**Issue by gsilvis** [CCI-MOC/haas#559]
This solves a lot of our use-cases.  What we want is the following API calls:

- Create VPN endpoint
- Delete VPN endpoint
- Attach VPN endpoint to network
- Detach VPN endpoint from network
- Get connection details

or something similar.  ("Create VPN endpoint for network ____" is another reasonable possibility.)

With this, people can run their own headnodes anywhere---for instance, on their personal laptop, or in a pre-existing VM host, such as Openstack.

Just like the headnode service, philosophically, it's better for this to be separate.  However, having this minimal version as an extension in HaaS seems very useful and easy.  Unlike the headnode service, it is easy to make it feature-complete enough for real use.

**Comment by gsilvis** [comment]
Thought:  I believe some switches may support doing VPN.  So we could have drivers that take advantage of that, so all the VPN traffic doesn't need to go through the HaaS master node

**Comment by henn** [comment]
> Thought: I believe some switches may support doing VPN. So we could
> have drivers that take advantage of that, so all the VPN traffic
> doesn't need to go through the HaaS master node
>
>

Great point! We should make the API implementation-agnostic.

Also - for the initial implementation, I had been thinking that we could
use lightweight containers (like docker) to isolate VPNs from each
other, other networks as well as the host.

**Comment by gsilvis** [comment]
I don't think that should be necessary---it should be possible for openvpn to route the packets directly to the relevant bridge, which will only have access to the relevant vlan

**Comment by zenhack** [comment]
@dschatzberg suggested containers too; I'm surprised how quickly you both reached for that. @gsilvis, is right, it's totally unnecessary. You can actually run a separate process per vpn config; the systemd service file on my laptop for openvpn is parametrized over the config file, so you can do e.g. `systemctl start openvpn@configname`. I think the right approach (at least for the openvpn backend, if we're going to make it pluggable), it to generate a config and then start a process per-vpn. If we want to support non-systemd systems we could use [supervisord](http://supervisord.org/). All of the examples I saw when I had a (quick) look had a cert for the client authentication; I don't know what else (if anything) openvpn supports, and I'm not super-familiar with the underlying protocols, but one thing to do if we wanted to use client certs is to have the create call (which will be a put or post) expect the client cert as part of the body of the request.

**Comment by dschatzberg** [comment]
Yeah, if it's just a matter of separate configs and you can simply run separate processes then we're good. FYI, I don't think a single openvpn process would be sufficient since you would need to restart it whenever you add/remove an endpoint. There are some advantages to using containers - clean separation of each config/keys/certificates and it makes it easy to distribute

There are basically two components to VPN, one is authentication (from client to server and vice versa) and the other is encryption (on the client <=> server path).

Authentication is all done through certificates. Each client and server will need a certificate signed by the same CA. Most deployments will just create their own CA for this which I think is good for HaaS too. The important thing here is that the client will need the ability to get a signed certificate from HaaS. Either they should be able to create their own and ask HaaS to sign it (and return the signed cert), or HaaS should create it and return the signed cert + key. It's not obvious to me if you want one CA per HaaS deployment, or one per vpn endpoint/server.

Encryption is done through one of two methods, either a shared, static TLS key (allowing one client, one server). I've set this up before and it is very easy, you would just allow users to query their endpoints key. The more involved, but flexible solution here would be running SSL. I have less of an idea of how SSL is configured here but you can probably use the same CA setup.

**Comment by henn** [comment]
Re: containers vs running on the host OS, the main advantage I see is
having defense in depth.

Imagine a client uploading an authorized public key to the VPN service,
which has to eventually be placed in some config file.
Would you rather that key be interpreted by an OpenVPN process that has
access to __every__ user network, betting the farm that there are no
Bobby Tables-type attacks, or have it interpreted by a process that only
has access to the network it needs access to and no root access to the
host (which is likely the haas master)?

Another possibility would be to have the server generate a symmetric or
asymmetric key for use (which gets rid of the need to pass user supplied
data), but if we have technology that keeps the server from needing to
need to know the credential in the first place why not use it?

**Comment by dschatzberg** [comment]
Containers don't help with the first issue. The most fine-grained permission you can give the openvpn container is CAP_NET_ADMIN since it needs to create taps, bridge them, etc. This permission effectively allows complete access to networks. If you wanted it finer grained you would need to use a VM. However, it should provide defense in depth in preventing disclosure of the private keys of each VPN server.

**Comment by zenhack** [comment]
> but if we have technology that keeps the server from needing to
> need to know the credential in the first place why not use it?

Spinning up whole containers is going to add a lot of implementation
complexity. This is a concern that always needs to be considered, and
I'm *deeply* uncomfortable with the amount of duct tape we're talking
about employing here.

This is starting to look complex enough that I'm not really comfortable
integrating all of the relevant moving parts into HaaS proper.

I'm also starting to think about other solutions, that don't use openvpn
as the underlying network technology. One possibility is to use Virtual
Distributed Eithernet:

http://vde.sourceforge.net/

We could rig up an instance of the `vde_plug` command to a websocket,
which would let us re-use all of our existing auth stuff -- no needing
to marshal credentials around, just open an HTTPS connection using your
usual credentials.

This would require tackling the same issues around thread use and tying
up a connection that came up with the console stuff, and I think in this
case we can't fudge it. Still, I'm more comfortable wrangling that than
all of the goop that seems to be involved in using openvpn.

**Comment by zenhack** [comment]
If we wanted the network to be a separate service, we could (instead of having HaaS do the websocket thing itself), just return a redirect to a capability URL on another service. @gsilvis, thoughts?

**Comment by zenhack** [comment]
Heck, we wouldn't even really need VDE, we could just open the tap and copy stuff.

**Comment by gsilvis** [comment]
I really don't think containers are necessary here.  I trust openvpn significantly more than I trust haas, to be honest, and haas already has access to the entire switch and to every VLAN the system (through the bridges)---never mind access to all of the headnodes which we're currently running.

If there's some way to get openvpn to reload its config without restarting, then we can use one big file.  If not, we can just run multiple openvpns.  As @zenhack pointed out, systemd makes it easy, and I'm sure there are other ways too.

**Comment by dschatzberg** [comment]
I agree that containers are unnecessary here. Can you explain how vde works? How would I DHCP over it? Also, how is this less complexity than VPN?

**Comment by zenhack** [comment]
I'm actually thinking VDE is an unnecessary component in what I'm proposing. Basically, creating and connecting to an endpoint would look like:

* Server creates a tap, attach it to the bridge.
* Client establishes a connection via a websocket. In case folks aren't familiar, this amounts to turning an HTTP connection into raw tcp, but has the advantage that you can do all of the usual header examination and policy stuff that you'd do with the HTTP request before starting the session.
* Server copies raw bytes between the websocket and the tap until the client disconnects.
* Client side tool exists for making the request and copying bytes between the connection and  its own tap.

So basically, just turn the HTTPS connection into an ethernet cable.

For authorization, we have the url be something like `/endpoint/${128 bits pulled out of /dev/urandom, in base64}`, and have HaaS convey this to the client.

As a bonus, I think with this scheme we can have this out of HaaS proper, and even loosely coupled enough that it's potentially useful and meaningful entirely outside of the context of HaaS. That would also mean we don't have to implement this stuff in python (maybe Go?), and maybe not have to deal with python's clumsy concurrency support, and the awkwardness that comes with trying to do this stuff inside of a wsgi app. Those details are up for debate, of course.

**Comment by zenhack** [comment]
This also means (if we rip out the headnodes and implement this as a separate service), HaaS doesn't *need* access to the bridges, or even need to run on the same machines.

**Comment by dschatzberg** [comment]
Can you elaborate a bit more on what clients would need to do (e.g. what tools exist)? I'm completely ignorant here. I'm not sure I see the advantage. It seems like you have more implementation complexity here because you actually have to implement the data path management (creating websockets, pulling data from them and copying to taps) versus just using openvpn. What python concurrency support do you have to handle? Isn't python just being used as a scripting language to setup VPNs in the original design?

**Comment by zenhack** [comment]
@dschatzberg, in the original desgin, yeah, python isn't an issue. With my proposal it's annoying (but probably doable), because wsgi assumes a normal-ish request/response pattern, and we'd have to know a bit more about the server. (wsgi is analogous to CGI, but python specific, if you're unfamiliar).

Things I'd like to avoid: 

* having write access to a directory in `/etc`, where openvpn stores its stuff
* having to mess with generating config files
* having to figure out how to get restricted access to systemd; enough to start/stop openvpn instances and not much else. not enough to be effectively root. I'm not even comfortable assuming a priori that you can write an openvpn config and not be able to get it to do arbitrary stuff as root
* having a hard dependency on systemd
* writing multiple backends for different daemon supervisors
* managing certs
* putting more logic in HaaS proper than we need to (I think we all agreed at the meeting that having this as part of HaaS core was at least philosophically a bit ugly, but it's also more stuff in a component that needs to be trusted).
* Having to set up all of that crap to install HaaS.

"Creating" a websocket doesn't involve much, and there are a good number of off-the-shelf libraries for what it does involve. You've already *got* a tcp connection, you just actually use it as one. Background reading:

https://en.wikipedia.org/wiki/WebSocket

With a library in place, it pretty much means at some point during the request handler we call something that says "make this a websocket" and we're done. The finicky thing is being able to kill the endpoint when it has active connections. Here's an implementation of the request handler in go-ish psuedocode, which doesn't handle asynchronously killing the connection (but hopefully gives an idea about the rest of the stuff):

```
Handle("/endpoint/<unguessable_id>", func(unguessable_id, request) {
    bridge := getBridgeNameFromDB(unguessable_id)
    if (!bridge) {
         // No bridge for the id; it's probably a bogus id.
         return 404
    }
    sock := websocket(request)
    registerConn(unguessable_id, sock)
    defer sock.Close()  // defer runs this when the function returns
    tap := makeTap(bridge)
    if !tap {
        return 500
    }
    defer tap.Delete()
    ch := chan()
    go func() { Copy(sock to tap); ch <- done }  // this is run in parallel
    Copy(tapConn to sock)
    <-ch // wait for other routine
})
```

There's stuff in the go standard library that could be used to make it safe to close those connections concurrently with the reads/writes, at which point this will just return. It's not totally trivial, but I don't think its worse than trying to pin down all of the policy constraints of the things we have to talk to (and estimate how likely bugs are) in all of the random crap we're going to have to interface with to do the other thing, particular when as ops folks we'd have to get a good chunk of that right several times, as a lot of it happens at install time. I'd rather deal with the slightly tricky thing once than have to deal with *even more* rube-goldberg machines when setting up my cloud.

**Comment by dschatzberg** [comment]
I think a lot of your concerns with the VPN solution would actually be resolved by a container/docker solution:

 - having write access to a directory in /etc, where openvpn stores its stuff
 - having to figure out how to get restricted access to systemd; enough to start/stop openvpn instances and not much else. not enough to be effectively root. I'm not even comfortable assuming a priori that you can write an openvpn config and not be able to get it to do arbitrary stuff as root
 - having a hard dependency on systemd
 - writing multiple backends for different daemon supervisors
 - putting more logic in HaaS proper than we need to (I think we all agreed at the meeting that having this as part of HaaS core was at least philosophically a bit ugly, but it's also more stuff in a component that needs to be trusted).
 - Having to set up all of that crap to install HaaS.

All of the above seem trivially solved by having a dockerfile which runs openvpn and can be used to generate keys, certs, and configs. Whats the issue with generating config files? A template that you modify/construct seems very straightforward (certainly easier to do and debug then to write your own application). Similarly, certs are setup once and done with. This seems like development investment that doesn't burden HaaS afterwards. It occurs to me that the biggest differences between websockets and VPN are the following:
- permission model (running openvpn in a container with CAP_NET_ADMIN vs running some custom software with CAP_NET_ADMIN too?... now that I write this, I don't see a significant difference)
- Client interaction: openvpn is installable on every major distro vs ?

I'm not sure what you are talking about in the last paragraph? It seems like setting up openvpn is the "deal with the slightly tricky thing once" solution. Once you've scripted up all your interactions/created your Dockerfiles it's done.

As a kibitzer on the project, I think the VPN solution is superior. From my perspective as a HaaS user, I could live with either solution.

**Comment by zenhack** [comment]
I basically don't trust containers with anything that can't be done as a regular user. The last time I checked, related vulnerabilities were *really* frequent. I'm having a hard time finding any positive confirmation that the situation has changed. 

Re: the client end, it's it's a trivial one-off program:

```
tap = openTap(argv[1])
sock = openWebsocket(argv[2])
go Copy(tap, sock)
Copy(sock, tap)
```

So both of those differences you mentioned are pretty much non-existant.

I guess the main thing is, I actually want to be able to understand all of this. I'm frankly fed up with the shove-the-tower-of-babel-in-a-container-and-forget-about-it approach. It seems like the rest of the industry has given up on writing software that's actually administrable by humans. I haven't. So far, HaaS is still something an intern can hope to set up manually. I want to keep it that way, and I don't think shoving the whole thing in a container counts any more than just telling people the only way to do it is to use the puppet manifests. I'm kindof offended by the notion of downloading an entire linux distro (and thinking about the implications of having to update that) to avoid writing a few hundered lines of code, at the most. I have an obvious bias, but I *do* think keeping the system understandable is important.

Honestly, I'm not sure you're wrong, but I don't think I want to be the one to implement it if that's the call we make.

**Comment by pjd-nu** [comment]
This whole discussion is making me very nervous. What are the particular issues with a straightforward implementation of one OpenVPN process per project, what I think was the original solution discussed? In particular:

1. use a regular (.deb or .rpm) installation of OpenVPN, but don't turn on its systemd.
2. create a config file (doesn't have to be in /etc) for each project
3. start an openvpn process for each project, pointing at its own config file

The biggest issue I see is the cert management, although there are a couple of other openvpn auth plugins that could be used. (e.g. the PAM plugin)

**Comment by zenhack** [comment]
@pjd-nu, if we're not going to use (something like) systemd, then we need to do process monitoring/shutdown somehow. HaaS's current architecture really isn't suited for it, again because of the assumed request/response control flow. The other issue is creating the tap devices, which can't be done as an unpriviledged user. We might be able to get away with pre-allocating those and setting the permissions such that the openvpn user can use them, but I'm not entirely sure, and that limits us to one (or a fixed number of) taps per bridge (maybe ok).

**Comment by pjd-nu** [comment]
We could create individual service files, again in a HaaS VPN-specific directory, and install them with systemctl link. (this would require running another copy of systemd as non-root: https://wiki.archlinux.org/index.php/Systemd/User)

**Comment by pjd-nu** [comment]
Well, it would require non-root systemd iff we wanted this whole thing to run non-root.

**Comment by zenhack** [comment]
Re: client certs, I think having the client generate them is a bad idea; all the examples I've looked at specify the path to the client cert in the config, which makes me distrust input validation of the client cert itself. It may be fine, but it doesn't seem like a well-trodden path. We could still have the server generate them, but do we really want to be having the server generate keypairs every time somewhat creates an endpoint? I don't know if `/dev/random` is up to the task.

I think I need to mull things over before I comment on anything else.

**Comment by gsilvis** [comment]
We should just use static-key mode. It's just two symmetric keys, which should be very easy to generate.  (They're not very long.)

**Comment by zenhack** [comment]
Quoting George Silvis, III (2016-03-28 23:09:13)

>    We should just use static-key mode. It's just two symmetric keys, which
>    should be very easy to generate. (They're not very long.)

Do you have a link describing that setup?

**Comment by gsilvis** [comment]
https://openvpn.net/index.php/open-source/documentation/miscellaneous/78-static-key-mini-howto.html

**Comment by henn** [comment]
RE: "Things I'd like to avoid", I think @dschatzberg
<https://github.com/dschatzberg> summarizes nicely how containers can
address the issues you raised.

RE: entropy, rngd (including in rng-tools) coupled with 'modprobe
tpm_rng' should do the trick.

**Comment by gsilvis** [comment]
Hard-depending on a full container solution like docker feels just as wrong to me as depending on the ability to run a user systemd session. User systemd sessions should do everything we need.

The openvpn config file is very easy to generate, too, and we already do that for things like that when using libvirt.

**Comment by gsilvis** [comment]
For that matter, none of this is a strict dependency, because you should clearly be able to disable this API in the config file.

**Comment by gsilvis** [comment]
@henn `/dev/random` (which is the random one---`/dev/urandom` is the pseudorandom one) for one symmetric key per haas API request should be fine.  If we don't use containers, there are many fewer cryptography things to get wrong.1

**Comment by zenhack** [comment]
I much prefer the user systemd session, which is not to say I like it. I don't have any better ideas though, and short of writing some custom thing (which I can't honestly say I think would be a good idea), I don't think I'm going to come up with any. Grumble.

Static key thing sounds good.

If we create the taps ahed of time (probably at the same time as the bridges) we can set the permissions such that we can just run openvpn as a regular user.

**Comment by dschatzberg** [comment]
Sounds good. Just to be clear, I don't think the static key resolves the need to manage certs, it just provides encryption from client to server (this is the 1 client, 1 server solution I was talking about via email). Yeah creating the taps for a particular user would let you run openvpn as that user. Are you planning to have one user per VPN or one for all?

**Comment by gsilvis** [comment]
@dschatzberg My understanding was that the static key contained both an encryption key and an authentication key, which would mean that no SSL certs are necessary.  The page I linked also claims "No X509 PKI (Public Key Infrastructure) to maintain"

**Comment by dschatzberg** [comment]
@gsilvis Hmm, I recall needing to generate a key when going down the cert route as well and figured this was a static key. Looking more closely at the description on the website, it seems you can just use the shared key without certs as you say. If so, that makes this very simple.

**Comment by zenhack** [comment]
My reading was the same as @gsilvis's; I think we can just generate the
key on the server side and give it to the client via one of the API
calls.

My thinking was to just run everything as the same user. Not doing that
would probably require *something* to be running as root, since we'd
need to dynamically spawn a process as a different user per network.

Anyway, it sounds like we have a decent outline. I can try to pin down a
concrete proposal/estimate.

**Comment by henn** [comment]
Apologies - somehow this got stuck in my outbox a week ago.

****

Other options to throw out there:

* [SSH's built-in Layer 2 VPN support using
taps](http://blog.rot13.org/2009/08/tap-magic-kvm-bridge-ssh-ethernet-tunnel.html)
  * May require letting users ssh in as root, though this could be
mitigated by the use of containers or VMs and limits within the
authorized_keys file
* Having haas make calls to a completely separate appliance (like
[IPFire](http://www.ipfire.org/)) that then sets things up.
